EventGenerator/src/GammaConversion_module.cc

```diff
@@ -0,0 +1,135 @@
+#include "Offline/Mu2eUtilities/inc/simParticleList.hh"
+#include "Offline/Mu2eUtilities/inc/GammaPairConversionSpectrum.hh"
+
+namespace mu2e {
+
+  //================================================================
+  class GammaConversion : public art::EDProducer {
+  public:
+    struct Config {
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+      fhicl::Atom<art::InputTag> inputSimParticles{Name("inputSimParticles"),Comment("A SimParticleCollection with input converted photons.")};
+      fhicl::Atom<int> material{Name("material"), Comment("Conversion material Z"), 13};
+      fhicl::Atom<bool> correlateAnglesOverKE{Name("correlateAnglesOverKE"), Comment("Option to correlate angles over KE"), true};
+      fhicl::Atom<int> verbosity{Name("verbosity"), Comment("Verbosity level"), 0};
+    };
+
+    using Parameters= art::EDProducer::Table<Config>;
+    explicit GammaConversion(const Parameters& conf);
+
+    virtual void produce(art::Event& event) override;
+
+    //----------------------------------------------------------------
+  private:
+
+    art::ProductToken<SimParticleCollection> const simsToken_;
+    const int material_;
+    const int verbosity_;
+
+    art::RandomNumberGenerator::base_engine_t& eng_;
+    CLHEP::RandFlat randFlat_;
+    GammaPairConversionSpectrum spectrum_;
+    GammaPairConversionSpectrum::elementData element_;
+    const ProcessCode process_;
+  };
+
+  //================================================================
+  GammaConversion::GammaConversion(const Parameters& conf)
+    : EDProducer{conf}
+    , simsToken_{consumes<SimParticleCollection>(conf().inputSimParticles())}
+    , material_{conf().material()}
+    , verbosity_{conf().verbosity()}
+    , eng_{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
+    , randFlat_{eng_}
+    , spectrum_(&randFlat_, conf().correlateAnglesOverKE())
+    , element_(spectrum_.GetElementMap()[material_])
+    , process_(ProcessCode::mu2eGammaConversion)
+  {
+    produces<mu2e::StageParticleCollection>();
+  }
+
+  //================================================================
+  void GammaConversion::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_);
+    const auto gammas = simParticleList(simh,PDGCode::gamma,ProcessCode::conv);
+
+    // If no suitable photons, print the particle collection and throw an error
+    if(gammas.empty()) {
+      const auto& sims = *simh;
+      std::cout << "GammaConversion_module::" << __func__ << ": Printing the particle collection:\n";
+      for(auto i = sims.begin(); i != sims.end(); ++i) {
+        const auto& inpart = i->second;
+        std::cout << "SimParticle: pdg = " << inpart.pdgId() << " createCode = " << inpart.creationCode()
+                  << " stopCode = " << inpart.stoppingCode() << std::endl;
+      }
+      throw cet::exception("BADINPUT")
```

Comment:
Why do you require the input collection always has a converted photon? Unless you filter the path leading to this module it seems impossible to guarantee. But maybe that's your intended workflow?

Reply:
I'll add a comment here, but the current workflow (that I'll make a PR for soon) produces output sim files that require a photon conversion in the event. So all input events should have a photon conversion. If an event doesn't we could pass an empty event as is done in the muon daughter resampling (`SingleProcessGenerator`, though `FlatMuonDaughterGenerator` does throw an error in this case), but this is likely the sign of a problem as we can't resample non-photon conversion events with this generator.